### ISSUE TITLE
[Contributors Welcome] Python/C++ API Parity: torch.nn modules

### ISSUE THREAD

**Issue by yf225** [pytorch/pytorch#25883]
As part of the Python/C++ API parity work, we would like to have the following `torch::nn` modules and utilities in C++ API:

## Containers
- [ ] Module (**TODO**: some APIs are missing in C++, e.g. `register_forward_hook`)
- [x] Sequential (@ShahriarSS)
- [x] ModuleList (@ShahriarSS)
- [ ] ModuleDict
- [ ] ParameterList
- [ ] ParameterDict

## Convolution layers
- [ ] Conv1d (@krshrimali for checking whether the current implementation matches the Python one)
- [ ] Conv2d
- [ ] Conv3d
- [ ] ConvTranspose1d
- [ ] ConvTranspose2d
- [ ] ConvTranspose3d
- [ ] Unfold
- [ ] Fold (@ShahriarSS https://github.com/pytorch/pytorch/pull/24160)

## Pooling layers
- [ ] MaxPool1d (@ShahriarSS https://github.com/pytorch/pytorch/pull/24860)
- [ ] MaxPool2d (@ShahriarSS https://github.com/pytorch/pytorch/pull/24860)
- [ ] MaxPool3d (@ShahriarSS https://github.com/pytorch/pytorch/pull/24860)
- [ ] MaxUnpool1d
- [ ] MaxUnpool2d
- [ ] MaxUnpool3d
- [ ] AvgPool1d (@ShahriarSS https://github.com/pytorch/pytorch/pull/25800)
- [ ] AvgPool2d (@ShahriarSS https://github.com/pytorch/pytorch/pull/25800)
- [ ] AvgPool3d (@ShahriarSS https://github.com/pytorch/pytorch/pull/25800)
- [ ] FractionalMaxPool2d
- [ ] LPPool1d
- [ ] LPPool2d
- [ ] AdaptiveMaxPool1d
- [ ] AdaptiveMaxPool2d
- [ ] AdaptiveMaxPool3d
- [ ] AdaptiveAvgPool1d
- [ ] AdaptiveAvgPool2d
- [ ] AdaptiveAvgPool3d

## Padding layers
- [ ] ReflectionPad1d
- [ ] ReflectionPad2d
- [ ] ReplicationPad1d
- [ ] ReplicationPad2d
- [ ] ReplicationPad3d
- [ ] ZeroPad2d
- [ ] ConstantPad1d
- [ ] ConstantPad2d
- [ ] ConstantPad3d

## Non-linear activations (weighted sum, nonlinearity)
- [ ] ELU
- [ ] Hardshrink
- [ ] Hardtanh
- [ ] LeakyReLU
- [ ] LogSigmoid
- [ ] MultiheadAttention
- [ ] PReLU
- [ ] ReLU
- [ ] ReLU6
- [ ] RReLU
- [ ] SELU
- [ ] CELU
- [ ] Sigmoid
- [ ] Softplus
- [ ] Softshrink
- [ ] Softsign
- [ ] Tanh
- [ ] Tanhshrink
- [ ] Threshold

## Non-linear activations (other)
- [ ] Softmin
- [ ] Softmax
- [ ] Softmax2d
- [ ] LogSoftmax
- [ ] AdaptiveLogSoftmaxWithLoss

## Normalization layers
- [ ] BatchNorm1d
- [ ] BatchNorm2d
- [ ] BatchNorm3d
- [ ] GroupNorm
- [ ] SyncBatchNorm
- [ ] InstanceNorm1d
- [ ] InstanceNorm2d
- [ ] InstanceNorm3d
- [ ] LayerNorm
- [ ] LocalResponseNorm

## Recurrent layers
- [ ] RNN
- [ ] LSTM
- [ ] GRU
- [ ] RNNCell
- [ ] LSTMCell
- [ ] GRUCell

## Transformer layers
NOTE: the following modules will be worked on by bootcamper
- [ ] Transformer
- [ ] TransformerEncoder
- [ ] TransformerDecoder
- [ ] TransformerEncoderLayer
- [ ] TransformerDecoderLayer

## Linear layers
- [ ] Identity
- [ ] Linear
- [ ] Bilinear
- [ ] Flatten

## Dropout layers
- [ ] Dropout
- [ ] Dropout2d
- [ ] Dropout3d
- [ ] AlphaDropout

## Sparse layers
NOTE: the following modules will be worked on by bootcamper
- [ ] Embedding
- [ ] EmbeddingBag

## Distance functions
- [ ] CosineSimilarity
- [ ] PairwiseDistance

## Loss functions
NOTE: @ShahriarSS is working on all of the following:
- [ ] L1Loss
- [ ] MSELoss
- [ ] CrossEntropyLoss
- [ ] CTCLoss
- [ ] NLLLoss
- [ ] PoissonNLLLoss
- [ ] KLDivLoss
- [ ] BCELoss
- [ ] BCEWithLogitsLoss
- [ ] MarginRankingLoss
- [ ] HingeEmbeddingLoss
- [ ] MultiLabelMarginLoss
- [ ] SmoothL1Loss
- [ ] SoftMarginLoss
- [ ] MultiLabelSoftMarginLoss
- [ ] CosineEmbeddingLoss
- [ ] MultiMarginLoss
- [ ] TripletMarginLoss

## Vision layers
- [ ] PixelShuffle
- [ ] Upsample
- [ ] UpsamplingNearest2d
- [ ] UpsamplingBilinear2d

## DataParallel layers (multi-GPU, distributed)
- [ ] DataParallel
- [ ] DistributedDataParallel

## Utilities
- [ ] clip_grad_norm_
- [ ] clip_grad_value_
- [ ] parameters_to_vector
- [ ] vector_to_parameters
- [ ] weight_norm
- [ ] remove_weight_norm
- [ ] spectral_norm
- [ ] remove_spectral_norm
- [ ] PackedSequence
- [ ] pack_padded_sequence
- [ ] pad_packed_sequence
- [ ] pad_sequence
- [ ] pack_sequence

## Implementation Notes:
- For `torch.nn` modules, Python and C++ implementation must only differ in language-specific syntax, and their data member fields, control flow and logic must be **exactly** the same.
- Besides ad-hoc tests to make sure the C++ `torch::nn` modules behave as expected, we also have `test/test_cpp_api_parity.py` which is a generic way to test parity of the following:
    - Module constructor args
    - Module state after construction
    - Output of forward given the same input
    - Parameter gradients computed by backward given the same input and the same output gradient.
- When you open a PR to add a C++ `torch::nn` module, please ask for @yf225's review, and I will add the appropriate parity test entries in `test/test_cpp_api_parity.py` for your new module.

**Comment by nmerrill67** [comment]
ReLU6 is of particular interest I think in the C++ API since it is useful for embedded platforms. As of now I cannot find a way to access the functional form relu6 either.

**Comment by JoshVarty** [comment]
@yf225 

We should figure out a good way to assign these issues. A couple of days ago I saw the [discussion](https://discuss.pytorch.org/t/contributors-welcome-add-missing-torch-nn-layers-in-c-api/55650)  thread and [I mentioned](https://discuss.pytorch.org/t/contributors-welcome-add-missing-torch-nn-layers-in-c-api/55650/23?u=joshvarty) I was interested in `ReflectionPad1d` and `ReflectionPad2d` and didn't see any mention of them being claimed in this thread (they were apparently edited in yesterday).

I realize now I should have also checked for existing PRs as they've been implemented in #26602, but it would be helpful if we could figure out a good way to assign or claim these work items so we don't simultaneously implement the same features.

I'm still interested in helping out. Is the list as it stands in this issue accurate? Can I claim a different feature to work on?

**Comment by yf225** [comment]
@JoshVarty 

Thanks a lot and really appreciated your interest in helping, and my sincere apologies for not updating the list early enough, which caused the work conflict. From now on, I will make sure to put work-in-progress items on this list as soon as I am aware of the effort, and please reply to this issue for the feature you are interested in working on and I will mention your name next to the feature.

The reason I didn't start asking other community members to take a look at this issue and claim their interested modules is because there are still a lot of preparation work that needs to be done (e.g. fixing how we handle enums in C++ API https://github.com/pytorch/pytorch/issues/15149). The community members also [asked for example PRs](https://discuss.pytorch.org/t/contributors-welcome-add-missing-torch-nn-layers-in-c-api/55650/15?u=yf225), which is the main reason why we proactively work on some modules in the first place (so that those PRs can serve as examples). Without those preparation work and example PRs, the experience of implementing new modules might be a bit more rocky for newcomers, but I don't think there would be any concerns for you. :)

Thanks again for your interest! Please let me know. :)

**Comment by yf225** [comment]
@nmerrill67 Thanks for the assessment and agreed that having ReLU6 would be very useful. Curious would you be interested in contributing the module `ReLU6` and/or the functional form `relu6`? Thanks a lot!

**Comment by nmerrill67** [comment]
@yf225 I will sure give it a try. I will have to spend some time looking at other examples that did this, as I could not figure it out immediately.

**Comment by yf225** [comment]
@nmerrill67 It seems that the Python version `ReLU6` is a subclass of `Hardtanh`, so I think probably the easiest way is to wait for https://github.com/pytorch/pytorch/pull/27038 to be merged, and then C++ `ReLU6` can just subclass from `Hardtanh` and pass in specific parameters like how the Python version does it.

cc. @pbelevich

**Comment by PyExtreme** [comment]
@yf225 , I am new to PyTorch. Could you please help me pick one so that I can get familiar.

Also, I have completed my work on `TripletMarginLoss` and would be reviewing for the final time today before making a PR.

Thanks

**Comment by PyExtreme** [comment]
@yf225 , I am planning to work on CrossEntropyLoss and NLLLoss. Please update it in the list.

Thank you

**Comment by yf225** [comment]
@PyExtreme Thanks so much for your help! I updated the list accordingly.

**Comment by nuka137** [comment]
@yf225 

I'm now tackling `LPPool1d` and `LPPool2d` while the Softmax related APIs are reviewing.
Could you update them in the list? Thanks

**Comment by yf225** [comment]
@nuka137 Just updated the list. Thanks so much for your help!

**Comment by jon-tow** [comment]
@yf225, I'd love to work on `Upsample` if that's okay!

**Comment by yf225** [comment]
@jon-tow Yes definitely! Thanks so much for your help :D

**Comment by CarMiranda** [comment]
Hi @yf225, I have started working on PixelShuffle and would also like to take on MultiLabelMarginLoss, SmoothL1Loss, SoftMarginLoss, MultiLabelSoftMarginLoss! Would that be okay?

**Comment by yf225** [comment]
@CarMiranda Yes definitely! I've just marked them on the list. Thanks so much for the great work you have done / are working on!

**Comment by Suyash458** [comment]
Hi @yf225 I'd like to pick up one of the four available loss functions. Which one do you suggest for a beginner?

**Comment by yf225** [comment]
@Suyash458 Thanks a lot for your interest! It seems that all loss functions are taken now. :/ If you are also interested in working on other layers, I think `AlphaDropout` would be a good layer to start with. Please let me know :D

**Comment by Suyash458** [comment]
@yf225 Thanks! That sounds good, I'll start with `AlphaDropout`

**Comment by gtamba** [comment]
@yf225 I'd like to take on Conv1D, Conv2D and Conv3D , let me know if I can open up issues for them.

**Comment by yf225** [comment]
@gtamba Thanks so much for your interest in working on the Conv layers! Currently the Conv layers are blocked by https://github.com/pytorch/pytorch/pull/26601, which is in turn blocked by https://github.com/pytorch/pytorch/pull/26837. Once the enum changes are made and `F::pad` is added, we will be able to work on the Conv layers. (And please let me know if you would still be interested in working on them at that time :D)

**Comment by gtamba** [comment]
Ah I see, of course I'd like to give it a stab at a later date as well, is there anything else that I could pick up for now? Like some of the Dropout modules?

**Comment by yf225** [comment]
@gtamba Thanks! Yes I think the Dropout modules need an overhaul, specifically I think the scope of work is as follows:
1. We want to templatize `DropoutImplBase` to work on all of 1D / 2D / 3D, similar to the design of `AvgPool` and `MaxPool` (in `torch/csrc/api/include/torch/nn/modules/pooling.h`). The 1D layer should still be named `Dropout`, the 2D layer should be named `Dropout2d`, and the 3D layer should be named `Dropout3d`, to all match the Python version.
2. Currently we have the 1D Dropout module (named `Dropout`) in the C++ API codebase, but it only takes an option `rate` instead of both `p` and `inplace` like the Python version, and we want to change it to match the Python version. Same for `Dropout2d` and `Dropout3d` as well. (`Dropout2dOptions` can just alias `DropoutOptions`, similar to how we do it for `AvgPool2dOptions` in `torch/csrc/api/include/torch/nn/options/pooling.h`. Same for `Dropout3dOptions` as well.)
3. The `FeatureDropout` layer that currently exists in the codebase should be deprecated in favor of the newly added `Dropout2d` and `Dropout3d`. We should print a deprecation warning in the constructor of `FeatureDropoutImpl` telling people to switch to either `Dropout2d` or `Dropout3d` (we should suggest the right alternative based on the shape of their input).

Thanks so much again and apologies for the large scope of work - C++ API unfortunately already has legacy code which we will hopefully get rid of one day and achieve 100% parity with Python API  :)

**Comment by mrsalehi** [comment]
Hi @yf225. I would like to take one of the layers. Which one do you suggest for a beginner?

**Comment by nuka137** [comment]
@yf225 

Hi, I almost finished to implement LPPool.
Could you assign `BatchNorm{1,2,3}` while reviewing?

**Comment by yf225** [comment]
@mrsalehi Thanks so much for your interest! I think `Flatten` would be a great layer to start with. Please let me know. :D

**Comment by yf225** [comment]
@nuka137 Thank you so much and really appreciated your help! I just mentioned your name next to  `BatchNorm{1,2,3}d`, and I think concretely we would need to do the following:
1. Currently there is already a `BatchNorm` layer in the C++ API, and we should deprecate it by showing a warning message in its constructor.
2. We should add `BatchNorm{1,2,3}d` layers by templatizing a `BatchNormImplBase` class, similar to the design of `AvgPool` and `MaxPool` in `torch/csrc/api/include/torch/nn/modules/pooling.h`. However I think we should implement the modules' `forward` / `reset_running_stats` / `reset_parameters` in the `BatchNormImplBase` class, to strictly follow the Python version.
3. We would need to modify `BatchNormOptions` to allow exactly the same set of arguments as the Python version (e.g. as part of the change, `features` needs to be renamed to `num_features`, and we need to add `track_running_stats`).

Thanks so much for your help again! The scope of work for `BatchNorm{1,2,3}d` is relatively larger because the Python version is a bit more complex than the other layers. And it would be a fantastic addition to the C++ API since so many people use it. :D

**Comment by mrsalehi** [comment]
@yf225 Yes, I would like to work on the `Flatten` layer!

**Comment by yf225** [comment]
@mrsalehi Thanks so much for your help! :D

**Comment by mailcorahul** [comment]
hi @yf225. I would like to work on any of the Normalization or Vision layers. Kindly suggest me one!

**Comment by yf225** [comment]
@mailcorahul Thanks a lot for your interest! I think `GroupNorm` would be a great one to start with. Please let me know if you would like to work on it :D

**Comment by mailcorahul** [comment]
Sure @yf225 , I can work on GroupNorm!

**Comment by yf225** [comment]
@mailcorahul Thanks a lot! I just mentioned your name next to it :D

**Comment by mansoorcheema** [comment]
Hi @yf225 I am interested to work on any Normalization Layer. I am beginner, so which one would you suggest. Thanks

**Comment by yf225** [comment]
@mansoorcheema Thanks a lot for your interest! I think `LocalResponseNorm` would be a great one to start with. Let me know if you are interested in working on it :D

**Comment by mansoorcheema** [comment]
@yf225 sure, I shall work on LocalResponseNorm!

**Comment by CarMiranda** [comment]
@yf225 since the loss PRs are (almost) done, and the pixel shuffle one is right around the corner, I'd like to work on the recurrent layers!

**Comment by yf225** [comment]
@mansoorcheema Thanks a lot! I just mentioned your name next to it :D

**Comment by yf225** [comment]
@CarMiranda Thanks a lot! I think the RNN layers are due for a rewrite as well to make sure they maintain parity with the Python version, but I think we can tackle the cells first. Thanks again for your help!

**Comment by divyanshsinghvi** [comment]
@yf225 I am interested to work on Normalization Layers. Could you suggest me one to start with?

**Comment by yf225** [comment]
@divyanshsinghvi Thanks a lot for your interest! Would you be interested in working on InstanceNorm1d / 2d / 3d? We would need to follow the design of AvgPool / MaxPool in `torch/csrc/api/include/torch/nn/modules/pooling.h` and templatize over a `InstanceNormImpl` base implementation. Also the modules / functionals / options should live in `instancenorm.h` / `.cpp` to reflect the file organization of the Python version. Thanks a lot for your help!

**Comment by divyanshsinghvi** [comment]
> @divyanshsinghvi Thanks a lot for your interest! Would you be interested in working on InstanceNorm1d / 2d / 3d? We would need to follow the design of AvgPool / MaxPool in `torch/csrc/api/include/torch/nn/modules/pooling.h` and templatize over a `InstanceNormImpl` base implementation. Also the modules / functionals / options should live in `instancenorm.h` / `.cpp` to reflect the file organization of the Python version. Thanks a lot for your help!

Sure. I will work on it.

**Comment by yf225** [comment]
@divyanshsinghvi Thanks! I mentioned your name next to the InstanceNorm layers :D

**Comment by mansoorcheema** [comment]
Hi @yf225 , I am working on LocalResponseNorm layer. I see in the python api that that i need pad function along with avg_pool2d/3d from functional but Pad and PadOptions are not available yet in the pytorch master branch. You can have a look at the proposed implementation in my commit here [https://github.com/mansoorcheema/pytorch/commit/76d8bdeb86ab00dc0311ea9c64382e27c12a27aa](url)

**Comment by nuka137** [comment]
@yf225 

Hi, I almost done `BatchNorm1d`.
If no one tackle to `Dropout{1,2,3}d`, I would like to tackle these modules.

**Comment by nsrishankar** [comment]
@yf225  Are the Conv layers/Losses still available?

**Comment by yf225** [comment]
@mansoorcheema Thanks so much for the commit and my apologies for the work-in-progress status of `pad` (it's currently blocked by https://github.com/pytorch/pytorch/pull/26837). I will review the commit tomorrow :D

**Comment by yf225** [comment]
@nuka137 Thanks a lot for your help! I think @gtamba is tackling `Dropout{1,2,3}d` (my apologies that I forgot to update the list). Curious would you be interested in working on `ConvTranspose{1, 2, 3}d`?

**Comment by yf225** [comment]
@nsrishankar Thanks a lot for your interest! I think the Conv layers and Losses are already taken / potentially taken by someone, wondering would you be interested in working on one of the following?
```
FractionalMaxPool2d
SyncBatchNorm
CrossMapLRN2d
GLU
AdaptiveLogSoftmaxWithLoss
```
Thanks a lot again for your help!

**Comment by nuka137** [comment]
@yf225 

Thanks for suggestion.
`ConvTranspose{1, 2, 3}d` are interesting to me. I'll tackle them.

**Comment by yf225** [comment]
@nuka137 Thanks a lot! I just mentioned your name next to them :D

**Comment by mansoorcheema** [comment]
@yf225 Thanks for the update. Meanwhile I would like to work on AdaptiveLogSoftmaxWithLoss.

**Comment by nsrishankar** [comment]
> @nsrishankar Thanks a lot for your interest! I think the Conv layers and Losses are already taken / potentially taken by someone, wondering would you be interested in working on one of the following?
> 
> ```
> FractionalMaxPool2d
> SyncBatchNorm
> CrossMapLRN2d
> GLU
> AdaptiveLogSoftmaxWithLoss
> ```
> 
> Thanks a lot again for your help!

@yf225 Could I take a crack at GLU and FractionalMaxPool2d if possible?

**Comment by lsrock1** [comment]
@yf225 Is grid_sample stiil available? If so, i want to tackle it.

**Comment by yf225** [comment]
@nsrishankar Definitely! I mentioned your name next to them. Thanks so much for your help! :D

**Comment by yf225** [comment]
@lsrock1 Thanks a lot for your help! Yes it's available and I just mentioned your name next to it :D

**Comment by jokerkeny** [comment]
Hi @yf225 , I want to give UpsamplingNearest2d a try, as I did superresolution before and I am really interested in this part. It's my first time contribute C++ in such a big project. Could I?

**Comment by PyExtreme** [comment]
Hi @yf225 , I have been working on `CrossEntropyLoss` and since there is no C++ Aten implementation of it, I have planned to write the code for the functional module's `loss.h` (_torch/csrc/api/include/torch/nn/functional/loss.h_) which doesn't rely on C/C++ either.
However, the formula of `CrossEntropyLoss` is not clear here in the docs:
https://pytorch.org/docs/stable/nn.html#crossentropyloss 
The terms in the formula have not been described. Could you please have a look into this and guide me the next steps?

Thank you

**Comment by yf225** [comment]
@jokerkeny Thanks a lot for your interest! I just found that `UpsamplingNearest2d` subclasses from `Upsample` and therefore we'll need to have `Upsample` first. @jon-tow Curious are you working on `Upsample`, or do you mind handing it over to @jokerkeny ? Thanks a lot! :D

**Comment by yf225** [comment]
@PyExtreme Thanks a lot for the investigation! I believe the `CrossEntropyLoss` module calls `F::cross_entropy()`, and in the Python version, `F.cross_entropy()` does the following:
```python
return nll_loss(log_softmax(input, 1), target, weight, None, ignore_index, None, reduction)
```
We already have `F::log_softmax` in C++ functional modules, but we still need to implement `F::nll_loss`. Therefore I believe the best way to sequence the work is for us to work on `torch::nn::NLLLoss` first (which calls `F::nll_loss` and as a result we need to work on that as a dependency), and then we can come back to `CrossEntropyLoss` to work on it.

**Comment by PyExtreme** [comment]
Hi @yf225 , yeah planning to translate that only but first work on `NLLLoss`.

**Comment by yf225** [comment]
@PyExtreme Thanks a lot for your help!

**Comment by jon-tow** [comment]
Hey @yf225! I’ve been working on `Upsample` but it’s been a bit awkward to port to C++. The Python API uses optional `size` and `scale_factor` List parameters for {1, 2, 3}D based inputs. The closest dynamic behavior would be using a `c10::optional<std::vector<...>>` for the two args but this leads to  unergonomic `UpsampleOptions` construction. For example, to provide a `scale_factor` option one would have to explicitly pass a temporary vector: `UpsampleOptions().scale_factor(std::vector<double>{0.5})`. Unfortunately,`ExpandingArray<D>` is off limits because we’d need to templatize and split `Upsample` which would break parity (think `Maxpool1D`, `MaxPool2D`, etc.).  
Do you have any suggestions for handling the optional `size` and `scale_factor` parameters? Thanks!

The recent docs mention that `UpsamplingNearest2D` and `UpsamplingBilinear2D` will be deprecated. Otherwise, they fall in easily once `Upsample` is complete!

**Comment by yf225** [comment]
@jon-tow Thanks so much for the investigation! Yes I think in that case we do need to use `c10::optional<std::vector<...>>` for `size` and `scale_factor` - I am wondering would `UpsampleOptions().scale_factor({0.5})` work or does it give an error? 

Thanks a lot for the catch about the deprecation of `UpsamplingNearest2D` and `UpsamplingBilinear2D` as well! @jokerkeny My apologies for the confusion and I think we can use @jon-tow's work on `Upsample` or `F::interpolate` whenever we need to use `UpsamplingNearest2d` or `UpsamplingBilinear2D` in the model, as the latter two are already deprecated in the Python API (I updated the list to reflect this). Please let me know if you would be interested in working on any other layers / functions, and thanks so much for your interest again!

**Comment by jokerkeny** [comment]
@yf225 That's ok. It's my bad that I didn't check the newest python api source code, where `UpsamplingNearest2D` and `UpsamplingBilinear2D` just call its parent `Upsample`'s method and depreciated.  

Can I try `clip_grad_value_` in Utilities?

**Comment by yf225** [comment]
@jokerkeney No it's my fault - my apologies that I didn't check them close enough when I wrote the list. :(  Yes it sounds awesome and thanks so much for your help! Please let me know if I can help in any way :D

**Comment by yf225** [comment]
@jokerkeny No it's my fault - my apologies that I didn't check them close enough when I wrote the list. :(  Yes it sounds awesome and thanks so much for your help! Please let me know if I can help in any way :D

**Comment by jon-tow** [comment]
@yf225 Constructing `scale_factor` and `size` with a single initializer-list of values causes an error. Because of the `c10::optional` wrapper one would need to nest the list, e.g. 
`UpsampleOptions().scale_factor({{0.5}})`.
I'll submit a WIP PR and maybe we could discuss it more there? Thanks for the help :)

**Comment by yf225** [comment]
@jon-tow Ah I see, yes we can discuss it in the WIP PR - I think there might be a way to get around this problem by explicitly having two function overloads for `.scalar_factor`, one takes `c10::nullopt_t` and the other takes `std::vector<double>`, but we might need to experiment it to see if it works :D

**Comment by Suyash458** [comment]
@yf225 The `AlphaDropout` module doesn't have an `inplace` parameter. Do we plan to add that to the Python/C++ API? For now, I've extended `DropoutImplBase` and added a functional/module test

**Comment by yf225** [comment]
@divyanshsinghvi Curious are you currently working on `InstanceNorm{1,2,3}d`, and is there anything I can do to help unblocking? Thanks!

**Comment by amn5277** [comment]
hi there i am new to opensource and would love to contribute to pytorch can any body suggest on how can i  help

**Comment by lsrock1** [comment]
@yf225 Can i tackle ParameterList and ParameterDict?

**Comment by divyanshsinghvi** [comment]
@yf225 yes I am working on the issue. I was trying to understand the structure of the code. I will get back to you with my initial draft in a 2-3 days. Will it be fine?

**Comment by yf225** [comment]
@lsrock1 Thanks a lot for your interest! I think one of our teammates is going to work on ParameterList and ParameterDict - curious would you be interested in working on `CrossMapLRN2d`? Thanks!

**Comment by yf225** [comment]
@divyanshsinghvi Yes that sounds awesome, thanks so much for your help! :D

**Comment by lsrock1** [comment]
@yf225  i am interested in CrossMapLRN2d, parameters_to_vector and vector_to_parameters. Is it possible?

**Comment by yf225** [comment]
@lsrock1 Thanks so much for your interest and definitely! I just mentioned your name next to the layer and the functions :D

And thanks a lot for the investigation! Yes I think we will need to write the custom backward and forward functions for `CrossMapLRN2d`. Currently we don't have a tutorial for writing custom Functions in C++, but I think we can use https://github.com/pytorch/pytorch/blob/dbf1996f796d027d88d5493cea1f5e4e3145d60c/test/cpp/api/autograd.cpp#L126-L152 and other tests in that file that subclasses from `public Function` as examples. Please let me know I can help clarify its usage in any way :D

**Comment by mansoorcheema** [comment]
Hi @yf225 I have completed the implementation code for LocalResponseNorm #28759  and wrote some tests. Can you have a look, specially at the tests. By looking at module tests for other modules, It seems that the output from the module is compared with the output from respective native function in aten, but for LocalResponseNorm there is no direct call to underlying native functions. For functional tests , for simple functions the output is compared with direct computation, while for some its compared with hardcoded expected output so i followed the latter approach. Let me know what you think.

**Comment by yf225** [comment]
> Hi @yf225 I have completed the implementation code for LocalResponseNorm #28759 and wrote some tests. Can you have a look, specially at the tests. By looking at module tests for other modules, It seems that the output from the module is compared with the output from respective native function in aten, but for LocalResponseNorm there is no direct call to underlying native functions. For functional tests , for simple functions the output is compared with direct computation, while for some its compared with hardcoded expected output so i followed the latter approach. Let me know what you think.

@mansoorcheema Thanks so much for the awesome work! I commented on the PR :D

**Comment by bernsm3** [comment]
@yf225 Can I take `weight_norm` and `remove_weight_norm`? Looking for a good first issue.

**Comment by yf225** [comment]
@bernsm3 Yes definitely! I just mentioned your name next to it. Thanks so much for your help!

**Comment by gtamba** [comment]
@yf225  : sorry it has taken so long, but I think I am done with the initial work for the Dropout modules, though there would be a merge conflict with #28424 , should I go ahead and open up a WIP PR?

**Comment by Yunseong-Jeong** [comment]
Can I join in?

**Comment by yf225** [comment]
> Can I join in?

@Yunseong-Jeong Thanks a lot for your interest! Would you like to work on any of the following functions?
```
spectral_norm
remove_spectral_norm
PackedSequence
pack_padded_sequence
pad_packed_sequence
pad_sequence
pack_sequence
```
Thanks so much for your help!

**Comment by yf225** [comment]
@bernsm3 Curious are you currently working on `weight_norm` and `remove_weight_norm`? Please let me know if there is anything I can help with :)

**Comment by Yunseong-Jeong** [comment]
@yf225 I'll try the spectral_norm. Is there anything I need to know before I do?

**Comment by Suyash458** [comment]
> 
> ```
> spectral_norm
> remove_spectral_norm
> PackedSequence
> pack_padded_sequence
> pad_packed_sequence
> pad_sequence
> pack_sequence
> ```
@yf225 I'd like to work on `PackedSequence`. Is that one still available?

**Comment by yf225** [comment]
> > ```
> > spectral_norm
> > remove_spectral_norm
> > PackedSequence
> > pack_padded_sequence
> > pad_packed_sequence
> > pad_sequence
> > pack_sequence
> > ```
> 
> @yf225 I'd like to work on `PackedSequence`. Is that one still available?

@Suyash458 Thanks so much for your interest! I just went through the list again and I realized that the RNN module workstream lead by @pbelevich would probably include `PackedSequence` as well. Among the rest of the functions, wondering would you be interested in working on `pad_sequence`? Thanks a lot again!

**Comment by yf225** [comment]
> @yf225 I'll try the spectral_norm. Is there anything I need to know before I do?

@Yunseong-Jeong My sincere apologies for the confusion - it seems that `spectral_norm` is currently blocked by the fact that we don't have `_forward_pre_hooks` in C++ `torch::nn::Module` at the moment. I will work on adding that to `torch::nn::Module` and keep you updated on the progress :D

**Comment by yf225** [comment]
> @bernsm3 Curious are you currently working on `weight_norm` and `remove_weight_norm`? Please let me know if there is anything I can help with :)

@bernsm3 My apologies for the earlier message - it seems that both `weight_norm` and `remove_weight_norm` are blocked by the fact that we don't have `_forward_pre_hooks` in C++ `torch::nn::Module` at the moment. I will work on adding that to `torch::nn::Module` and keep you updated on the progress :D

**Comment by Suyash458** [comment]
@yf225 `pad_sequence` sounds good!

**Comment by yf225** [comment]
@Suyash458 Thanks so much for your help! I mentioned your name next to it :D

**Comment by Yunseong-Jeong** [comment]
Then, If Spectral_norm is open Could you tell me?

**Comment by yf225** [comment]
> Then, If Spectral_norm is open Could you tell me? If you could do that, I'd appreciate it.
> Or if there's anything else I can do, I'd like to help. Please Mention me :)

@Yunseong-Jeong Yup I am working on the module hooks now and will let you know as soon as I finished it. Thanks so much for your help!

**Comment by RithvikSharma** [comment]
@yf225, I'd like to work on `pack_sequence`. Is it available? _ 

> 
> ```
> spectral_norm
> remove_spectral_norm
> PackedSequence
> pack_padded_sequence
> pad_packed_sequence
> pad_sequence
> pack_sequence
> ```
> 
> Thanks so much for your help!

**Comment by Suyash458** [comment]
@yf225 Where should the tests for the utility functions like `pad_sequence` reside?

**Comment by yf225** [comment]
@Suyash458 It would be awesome to put them in test/cpp/api/nn_utils.cpp. Thanks so much for your help!

**Comment by yf225** [comment]
> @yf225, I'd like to work on `pack_sequence`. Is it available? _

We are going through some internal planning at the moment, and I will update on its availability at the earliest possible :D

**Comment by pandeykartikey** [comment]
Hi, I would like to work on PackedSequence. If it is available. @yf225

**Comment by yf225** [comment]
cc. @glaringlee

**Comment by haoyang77** [comment]
@yf225 Hi, I would like to help out. Do you think ParameterList/ParameterDict is a good start for a beginner?

**Comment by yyn19951228** [comment]
I would like to take ParameterDict, and ParameterList if possible.

**Comment by glaringlee** [comment]
> I would like to take ParameterDict, and ParameterList if possible.
Sure. I will review your Parameter Dict PR #40014. 
For ParameterList, @haoyang77 has a PR #36904, but stopped updating, you can continue working on that. Thanks a lot for your help.

**Comment by yyn19951228** [comment]
> > I would like to take ParameterDict, and ParameterList if possible.
> > Sure. I will review your Parameter Dict PR #40014.
> > For ParameterList, @haoyang77 has a PR #36904, but stopped updating, you can continue working on that. Thanks a lot for your help.

Sure, I could continue working on that.

**Comment by k-melnyk** [comment]
@yf225 Hi. I would like to work on spectral_norm and remove_spectral_norm if nobody works on those.

**Comment by glaringlee** [comment]
> @yf225 Hi. I would like to work on spectral_norm and remove_spectral_norm if nobody works on those.

@k-melnyk 
Thanks for reaching out. I am currently maintaining this page. 
Yep, you can work on them. 
We have no hook support at the moment, you need to add forward_pre_hook as well. 
For hook, you can take a look at https://github.com/pytorch/pytorch/issues/25888#issuecomment-651902570
We have some draft work from @yf225, but he switched area and not able to finish it, it would be great if you can work on this hook work and those draft work can be a reference for you.

**Comment by jamaliki** [comment]
@glaringlee Hi, is there anything outstanding here, looks like some of the PRs are open? If there is something to do here I would like to help as my first contribution.

**Comment by ShaswatPanda** [comment]
Hi, I'd like to work on Module(Containers) and implement the APIs that are missing in C++, e.g. register_forward_hook / register_forward_pre_hook

**Comment by bjaeger1** [comment]
Hi! 
is somebody working on the implementation of the "register_forward_hook" function? I would really need that function to extract feature maps on different levels from one model.
My current work-around is loading 2 models (which are trimmed at the needed level) with torch::jit::load, which results in in managing two models instead of just one
